src/lexer.coffee

```diff
@@ -1252,7 +1252,20 @@ REGEX = /// ^
 REGEX_FLAGS  = /^\w*/
 VALID_FLAGS  = /^(?!.*(.).*\1)[imguy]*$/
 
-HEREGEX      = /// ^(?: [^\\/#] | \\[\s\S] | /(?!//) | \#(?!\{) )* ///
+HEREGEX      = /// ^
+  (?:
+      # Skip `\`, `/`, `#` and `\s` characters.
+      [^\\/#\s]
+      # Match `\` followed by any character.
+    | \\[\s\S]
+      # Match any `/` except `///`.
+    | /(?!//)
+      # Match `#` which is not part of interpolation, e.g. `#{}`.
+    | \#(?!\{)
+      # Comments consume everything until the end of the line, including `///`.
+    | \s+(?:#(?!\{)[^\n]*)?
+  )*
+///
```

Comment:
_Thank you_ for commenting this. One style note: shouldn’t the `|` line continuers be at the end of the line, rather than the start of the next one? That’s how we have things elsewhere in the codebase, e.g. `if firstCondition and \n secondCondition`.

@zdenko and @lydell, why does a heregex require a space before a `#` comment? That seems to go against the general rules for comments.

Reply:
> shouldn’t the | line continuers be at the end of the line, rather than the start of the next one

I kinda followed the style from other heregex patterns in `lexer.coffee`   
Examples:
```coffeescript
HEREGEX_OMIT = ///
    ((?:\\\\)+)     # Consume (and preserve) an even number of backslashes.
  | \\(\s)          # Preserve escaped whitespace.
  | \s+(?:#.*)?     # Remove whitespace and comments.
///g

REGEX_INVALID_ESCAPE = ///
  ( (?:^|[^\\]) (?:\\\\)* )        # Make sure the escape isn’t escaped.
  \\ (
     ?: (0[0-7])                   # octal escape
      | (x(?![\da-fA-F]{2}).{0,2}) # hex escape
      | (u\{(?![\da-fA-F]{1,}\})[^}]*\}?) # unicode code point escape
      | (u(?!\{|[\da-fA-F]{4}).{0,4}) # unicode escape
  )
///
```

> why does a heregex require a space before a # comment?

`HEREGEX_OMIT`, which removes comments from the heregex, requires a space before `#`   
`\s+(?:#.*)?`.
We could add the rule for escaping `#` unless it's a comment, which will allow comments without the preceding space.
But, there are many valid heregex patterns without escaped `#`, even in the `lexer.coffee`:  
`STRING_DOUBLE  = /// ^(?: [^\\"#] | \\[\s\S] | \#(?!\{) )* ///`, e.g. first `#` in `[^\\"#]`
The same heregex can be changed to 
`STRING_DOUBLE  = /// ^(?: [^\\"#] | \\[\s\S] |#(?!\{) )* ///` and it would still be valid:

```coffeescript
STRING_DOUBLE  = /// ^(?: [^\\"#] | \\[\s\S] | \#(?!\{) )* ///
# STRING_DOUBLE = /^(?:[^\\"#]|\\[\s\S]|\#(?!\{))*/;

STRING_DOUBLE  = /// ^(?: [^\\"#] | \\[\s\S] |#(?!\{) )* ///
# STRING_DOUBLE = /^(?:[^\\"#]|\\[\s\S]|#(?!\{))*/;

```

So, these existing patterns wouldn't be valid anymore if we allow heregex comments without preceding space.

Reply:
> @zdenko and @lydell, why does a heregex require a space before a # comment? That seems to go against the general rules for comments.

What @zdenko said. In other words: Because that’s the way it’s always been. Changing that now would be a breaking change. Even CoffeeScript’s own source code would need to be updated! And it’s easy to see the original thinking: Using literal spaces in regexes is uncommon. This means that you can copy-paste regex and paste into your heregex without having to think about escaping any `#`s.